database/dbServices/shinx.api.js

```diff
@@ -40,14 +40,20 @@ module.exports = {
     },
     async switchShininessAndGet(id) {
         let shinx = await this.getShinx(id, ['user_id', 'shiny']);
-        return shinx.switchShininessAndGet();
+        let user = await this.getUser(id, ['user_id']);
+        let hasShinyCharm = await this.hasEventTrophy(user.user_id, 'Shiny Charm');
+        if (hasShinyCharm) {
+            return shinx.switchShininessAndGet();
+        } else {
+            throw new Error('User does not have the Shiny Charm');
+        }
     },
     async changeAutoFeed(id, mode) {
         let shinx = await this.getShinx(id, ['user_id', 'auto_feed']);
         return shinx.setAutoFeed(mode);
     },
     async addExperience(id, experience) {
-        let shinx = await this.getShinx(id, ['user_id', 'experience']);
+        let shinx = await this.getShinx(id, ['user_id', 'experience', 'level']);
```

Comment:
Level is determined from the EXP count instead of being stored seperately so it doesn't really make sense to pass that in the arguments.
Level is calculated here: https://github.com/Glazelf/NinigiBot/blob/4b1e535883475dbf3293e34b36c4e2660619d76c/util/shinx/getExpFromLevel.js